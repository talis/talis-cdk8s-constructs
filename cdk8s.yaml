language: typescript
imports:
  - k8s@1.30.0
