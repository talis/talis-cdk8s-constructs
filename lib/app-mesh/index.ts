export * from "./ingress-gateway-nlb";
export * from "./app-service";
