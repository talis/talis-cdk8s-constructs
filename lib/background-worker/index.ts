export * from "./background-worker-props";
export * from "./background-worker";
