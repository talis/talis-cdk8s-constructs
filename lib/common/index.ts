export * from "./container-props";
export * from "./env-util";
export * from "./name-util";
export * from "./statefulset-util";
export * from "./value-util";
export * from "./workload-props";
