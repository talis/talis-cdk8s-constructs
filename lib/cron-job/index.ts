export * from "./cron-job-props";
export * from "./cron-job";
