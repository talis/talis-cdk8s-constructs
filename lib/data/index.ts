export * from "./config-map";
export * from "./secret";
export * from "./factory";
