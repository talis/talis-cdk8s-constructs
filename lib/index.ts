export * from "./cron-job";
export * from "./background-worker";
export * from "./data";
export * from "./job";
export * from "./redis";
export * from "./talis-chart";
export * from "./web-service";
