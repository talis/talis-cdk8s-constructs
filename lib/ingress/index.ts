export * from "./ingress-props";
export * from "./ingress";
