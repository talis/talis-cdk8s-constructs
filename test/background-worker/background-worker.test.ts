import { Chart, Testing } from "cdk8s";
import { Quantity } from "../../imports/k8s";
import { BackgroundWorker, BackgroundWorkerProps } from "../../lib";

const requiredProps = {
  image: "talis/app:worker-v1",
  release: "v1",
  resources: {
    requests: {
      cpu: Quantity.fromString("100m"),
      memory: Quantity.fromString("100Mi"),
    },
  },
};

function synthBackgroundWorker(props: BackgroundWorkerProps = requiredProps) {
  const chart = Testing.chart();
  new BackgroundWorker(chart, "worker-test", props);
  const results = Testing.synth(chart);
  return results;
}

describe("BackgroundWorker", () => {
  describe("Props", () => {
    test("Minimal required props", () => {
      const chart = Testing.chart();
      new BackgroundWorker(chart, "worker-test", requiredProps);
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();
    });

    test("All the props", () => {
      const app = Testing.app();
      const chart = new Chart(app, "test", {
        namespace: "test",
        labels: {
          app: "my-app",
          environment: "test",
          region: "local",
        },
      });
      const selectorLabels = {
        app: "my-app",
        instance: "test",
      };
      new BackgroundWorker(chart, "worker-test", {
        ...requiredProps,
        selectorLabels,
        workingDir: "/some/path",
        command: ["/bin/sh", "-c", "echo hello"],
        args: ["--foo", "bar"],
        env: [{ name: "FOO", value: "bar" }],
        envFrom: [{ configMapRef: { name: "foo-config" } }],
        automountServiceAccountToken: true,
        imagePullPolicy: "Always",
        imagePullSecrets: [{ name: "foo-secret" }],
        priorityClassName: "high-priority",
        revisionHistoryLimit: 5,
        affinity: {
          podAntiAffinity: {
            preferredDuringSchedulingIgnoredDuringExecution: [
              {
                podAffinityTerm: {
                  labelSelector: {
                    matchLabels: selectorLabels,
                  },
                  topologyKey: "topology.kubernetes.io/zone",
                },
                weight: 100,
              },
            ],
          },
        },
        resources: {
          requests: {
            cpu: Quantity.fromNumber(0.1),
            memory: Quantity.fromString("100Mi"),
          },
          limits: {
            cpu: Quantity.fromNumber(1),
            memory: Quantity.fromString("1Gi"),
          },
        },
        securityContext: {
          runAsUser: 1000,
          runAsGroup: 1000,
          runAsNonRoot: true,
        },
        replicas: 4,
        terminationGracePeriodSeconds: 300,
        lifecycle: {
          postStart: {
            exec: {
              command: ["/bin/sh", "-c", "echo hello"],
            },
          },
          preStop: {
            exec: {
              command: ["/bin/sh", "-c", "echo goodbye"],
            },
          },
        },
        livenessProbe: {
          exec: {
            command: [
              "/bin/sh",
              "-c",
              "test $(stat -c %Y /tmp/live) -gt $(($(date +%s) - 60))",
            ],
          },
          initialDelaySeconds: 30,
          periodSeconds: 10,
          failureThreshold: 3,
          successThreshold: 1,
          timeoutSeconds: 2,
        },
        volumes: [
          {
            name: "tmp-dir",
            emptyDir: {},
          },
        ],
        volumeMounts: [
          {
            name: "tmp-dir",
            mountPath: "/tmp",
          },
        ],
        initContainers: [
          {
            name: "init-container",
            image: "busybox:1.35.0",
            command: ["/bin/sh", "-c", "echo hello"],
          },
        ],
      });
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();
    });

    test("Allows specifying no affinity", () => {
      const results = synthBackgroundWorker({
        ...requiredProps,
        affinity: undefined,
      });
      const deployment = results.find((obj) => obj.kind === "Deployment");
      expect(deployment).not.toHaveProperty("spec.template.spec.affinity");
    });

    test("Allows specifying custom logic to make affinity", () => {
      const results = synthBackgroundWorker({
        ...requiredProps,
        makeAffinity(matchLabels) {
          return {
            podAffinity: {
              requiredDuringSchedulingIgnoredDuringExecution: [
                {
                  labelSelector: {
                    matchExpressions: [
                      {
                        key: "role",
                        operator: "In",
                        values: [matchLabels.role],
                      },
                    ],
                  },
                  topologyKey: "kubernetes.io/hostname",
                },
              ],
            },
          };
        },
      });
      const deployment = results.find((obj) => obj.kind === "Deployment");
      expect(deployment).toHaveProperty("spec.template.spec.affinity");
      expect(deployment.spec.template.spec.affinity).toMatchSnapshot();
    });

    test("Allows returning no affinity", () => {
      const results = synthBackgroundWorker({
        ...requiredProps,
        makeAffinity() {
          return undefined;
        },
      });
      const deployment = results.find((obj) => obj.kind === "Deployment");
      expect(deployment).not.toHaveProperty("spec.template.spec.affinity");
    });
  });

  describe("Container name", () => {
    test("Default container name", () => {
      const results = synthBackgroundWorker();
      const deployment = results.find((obj) => obj.kind === "Deployment");
      expect(deployment).toHaveProperty(
        "spec.template.spec.containers[0].name",
        "app"
      );
    });

    test("Container name from chart's app label", () => {
      const app = Testing.app();
      const chart = new Chart(app, "test", {
        labels: {
          app: "from-chart",
        },
      });
      new BackgroundWorker(chart, "worker-test", requiredProps);
      const results = Testing.synth(chart);
      const deployment = results.find((obj) => obj.kind === "Deployment");
      expect(deployment).toHaveProperty(
        "spec.template.spec.containers[0].name",
        "from-chart"
      );
    });

    test("Container name from selector label", () => {
      const results = synthBackgroundWorker({
        ...requiredProps,
        selectorLabels: { app: "from-selector" },
      });
      const deployment = results.find((obj) => obj.kind === "Deployment");
      expect(deployment).toHaveProperty(
        "spec.template.spec.containers[0].name",
        "from-selector"
      );
    });

    test("Container name set explicitly", () => {
      const results = synthBackgroundWorker({
        ...requiredProps,
        containerName: "explicit-name",
      });
      const deployment = results.find((obj) => obj.kind === "Deployment");
      expect(deployment).toHaveProperty(
        "spec.template.spec.containers[0].name",
        "explicit-name"
      );
    });
  });

  describe("Custom stop signal", () => {
    test("Either stopSignal or lifecycle.preStop can be specified", () => {
      expect(() => {
        new BackgroundWorker(Testing.chart(), "worker-test", {
          ...requiredProps,
          stopSignal: "KILL",
          lifecycle: {
            preStop: {
              exec: {
                command: ["kill", "-9", "1"],
              },
            },
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test("Setting stopSignal creates a preStop hook", () => {
      const results = synthBackgroundWorker({
        ...requiredProps,
        stopSignal: "QUIT",
      });
      const container = results[0].spec.template.spec.containers[0];
      expect(container.lifecycle).toMatchSnapshot();
    });

    test("Merge stopSignal and postStart hook", () => {
      const results = synthBackgroundWorker({
        ...requiredProps,
        stopSignal: "QUIT",
        lifecycle: {
          postStart: {
            exec: {
              command: ["/bin/sh", "-c", "echo hello"],
            },
          },
        },
      });
      const container = results[0].spec.template.spec.containers[0];
      expect(container.lifecycle).toMatchSnapshot();
    });
  });
});
