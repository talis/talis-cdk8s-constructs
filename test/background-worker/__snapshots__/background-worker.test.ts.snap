// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundWorker Custom stop signal Either stopSignal or lifecycle.preStop can be specified 1`] = `"stopSignal and lifecycle.preStop are mutually exclusive"`;

exports[`BackgroundWorker Custom stop signal Merge stopSignal and postStart hook 1`] = `
Object {
  "postStart": Object {
    "exec": Object {
      "command": Array [
        "/bin/sh",
        "-c",
        "echo hello",
      ],
    },
  },
  "preStop": Object {
    "exec": Object {
      "command": Array [
        "/bin/sh",
        "-c",
        "kill -QUIT 1 && while kill -0 1; do sleep 1; done",
      ],
    },
  },
}
`;

exports[`BackgroundWorker Custom stop signal Setting stopSignal creates a preStop hook 1`] = `
Object {
  "preStop": Object {
    "exec": Object {
      "command": Array [
        "/bin/sh",
        "-c",
        "kill -QUIT 1 && while kill -0 1; do sleep 1; done",
      ],
    },
  },
}
`;

exports[`BackgroundWorker Props All the props 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app": "my-app",
        "environment": "test",
        "instance": "test",
        "region": "local",
        "release": "v1",
        "role": "worker",
      },
      "name": "test-worker-test-c860ab13",
      "namespace": "test",
    },
    "spec": Object {
      "replicas": 4,
      "revisionHistoryLimit": 5,
      "selector": Object {
        "matchLabels": Object {
          "app": "my-app",
          "instance": "test",
          "role": "worker",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "my-app",
            "environment": "test",
            "instance": "test",
            "region": "local",
            "release": "v1",
            "role": "worker",
          },
        },
        "spec": Object {
          "affinity": Object {
            "podAntiAffinity": Object {
              "preferredDuringSchedulingIgnoredDuringExecution": Array [
                Object {
                  "podAffinityTerm": Object {
                    "labelSelector": Object {
                      "matchLabels": Object {
                        "app": "my-app",
                        "instance": "test",
                      },
                    },
                    "topologyKey": "topology.kubernetes.io/zone",
                  },
                  "weight": 100,
                },
              ],
            },
          },
          "automountServiceAccountToken": true,
          "containers": Array [
            Object {
              "args": Array [
                "--foo",
                "bar",
              ],
              "command": Array [
                "/bin/sh",
                "-c",
                "echo hello",
              ],
              "env": Array [
                Object {
                  "name": "FOO",
                  "value": "bar",
                },
              ],
              "envFrom": Array [
                Object {
                  "configMapRef": Object {
                    "name": "foo-config",
                  },
                },
              ],
              "image": "talis/app:worker-v1",
              "imagePullPolicy": "Always",
              "lifecycle": Object {
                "postStart": Object {
                  "exec": Object {
                    "command": Array [
                      "/bin/sh",
                      "-c",
                      "echo hello",
                    ],
                  },
                },
                "preStop": Object {
                  "exec": Object {
                    "command": Array [
                      "/bin/sh",
                      "-c",
                      "echo goodbye",
                    ],
                  },
                },
              },
              "livenessProbe": Object {
                "exec": Object {
                  "command": Array [
                    "/bin/sh",
                    "-c",
                    "test $(stat -c %Y /tmp/live) -gt $(($(date +%s) - 60))",
                  ],
                },
                "failureThreshold": 3,
                "initialDelaySeconds": 30,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 2,
              },
              "name": "my-app",
              "resources": Object {
                "limits": Object {
                  "cpu": 1,
                  "memory": "1Gi",
                },
                "requests": Object {
                  "cpu": 0.1,
                  "memory": "100Mi",
                },
              },
              "securityContext": Object {
                "runAsGroup": 1000,
                "runAsNonRoot": true,
                "runAsUser": 1000,
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/tmp",
                  "name": "tmp-dir",
                },
              ],
              "workingDir": "/some/path",
            },
          ],
          "imagePullSecrets": Array [
            Object {
              "name": "foo-secret",
            },
          ],
          "initContainers": Array [
            Object {
              "command": Array [
                "/bin/sh",
                "-c",
                "echo hello",
              ],
              "image": "busybox:1.35.0",
              "name": "init-container",
            },
          ],
          "priorityClassName": "high-priority",
          "terminationGracePeriodSeconds": 300,
          "volumes": Array [
            Object {
              "emptyDir": Object {},
              "name": "tmp-dir",
            },
          ],
        },
      },
    },
  },
]
`;

exports[`BackgroundWorker Props Allows specifying custom logic to make affinity 1`] = `
Object {
  "podAffinity": Object {
    "requiredDuringSchedulingIgnoredDuringExecution": Array [
      Object {
        "labelSelector": Object {
          "matchExpressions": Array [
            Object {
              "key": "role",
              "operator": "In",
              "values": Array [
                "worker",
              ],
            },
          ],
        },
        "topologyKey": "kubernetes.io/hostname",
      },
    ],
  },
}
`;

exports[`BackgroundWorker Props Minimal required props 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "instance": "worker-test",
        "release": "v1",
        "role": "worker",
      },
      "name": "test-worker-test-c860ab13",
    },
    "spec": Object {
      "revisionHistoryLimit": 1,
      "selector": Object {
        "matchLabels": Object {
          "instance": "worker-test",
          "role": "worker",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "instance": "worker-test",
            "release": "v1",
            "role": "worker",
          },
        },
        "spec": Object {
          "affinity": Object {
            "podAntiAffinity": Object {
              "preferredDuringSchedulingIgnoredDuringExecution": Array [
                Object {
                  "podAffinityTerm": Object {
                    "labelSelector": Object {
                      "matchLabels": Object {
                        "instance": "worker-test",
                        "role": "worker",
                      },
                    },
                    "topologyKey": "topology.kubernetes.io/zone",
                  },
                  "weight": 100,
                },
              ],
            },
          },
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "image": "talis/app:worker-v1",
              "imagePullPolicy": "IfNotPresent",
              "name": "app",
              "resources": Object {
                "requests": Object {
                  "cpu": "100m",
                  "memory": "100Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
]
`;
